lib/checkunusedvar.cpp

```diff
@@ -1267,7 +1267,8 @@ void CheckUnusedVar::checkFunctionVariableUsage()
                 continue;
 
             FwdAnalysis fwdAnalysis(mTokenizer->isCPP(), mSettings->library);
-            if (fwdAnalysis.unusedValue(expr, start, scope->bodyEnd)) {
+            const Token* scopeEnd = expr->variable() ? getEndOfVarScope(expr, getVariables(expr)) : scope->bodyEnd;
```

Comment:
Checking `expr->variable()` before calling `getVariables` defeats the purpose of calling `getVariables`. You should be doing something like this:

```cpp
const Token* scopeEnd = getEndOfVarScope(expr, getVariables(expr));
if (!scopeEnd)
    scopeEnd = scope->bodyEnd;
```

Reply:
That causes my latest test case to fail. The `[` in `m[n]` gets flagged as unused, but the whole container `m` is used later. Maybe `[` shouldn't be checked at all?

Reply:
> Maybe [ shouldn't be checked at all?

What do you meant should be checked?

The problem might be that its no finding the last token correctly in `getEndOfVarScope`. It looks like `precedes(varScope->bodyEnd, endOfVarScope)` might be backwards in `getEndOfVarScope`. It seems like it should be `precedes(endOfVarScope, varScope->bodyEnd)`.

Reply:
Ok, that seems to fix it.
I was just wondering how elements could be checked separately from their container.

Reply:
Now there's an FP for
~~~~
struct S { int i; };
S f(S s, bool b) {
    if (b)
        s.i = 1;
    return s;
}
~~~~

Reply:
According to `getEndOfVarScope()`, `i` goes out of scope at the end of `f()`. `s` has global scope, but `varScope->bodyEnd` is NULL.

Reply:
> i goes out of scope at the end of f(). s has global scope

Wouldn't `s` have scope to the end of `f()` and `i` would have scope to the send of `struct S`.

> but varScope->bodyEnd is NULL.

We probably need to check for that in `getEndOfVarScope`.

Reply:
> Wouldn't `s` have scope to the end of `f()` and `i` would have scope to the send of `struct S`.

I misread that, `i` gets an end token that points to the end of the `if` block.
> > but varScope->bodyEnd is NULL.
> 
> We probably need to check for that in `getEndOfVarScope`.

Is it ok to return NULL in that case?
What is `getEndOfVarScope` actually supposed to return for multiple variables? The end of the (largest/smallest/...) scope?

Reply:
> I misread that, i gets an end token that points to the end of the if block.

I see why now, it uses `tok->scope()` instead of `var->scope()` for non-local variables. Unfortunately, it uses `var->typeStartToken()->scope()` for local variables, which is why you get global scope for `S`. It should use `var->scope()` instead.

> Is it ok to return NULL in that case?

If its global scope then thats fine. Although, `S` is not global-scope. 

> What is getEndOfVarScope actually supposed to return for multiple variables? The end of the (largest/smallest/...) scope?

Well, it originally it returned the smallest, but for it would be better to return the largest scope especially for unreadVariables.

Reply:
Hmm, replacing `var->typeStartToken()->scope()` with `var->scope()` causes TestAutoVariables::invalidLifetime to fail.

Reply:
It seems to me that TestAutoVariables::invalidLifetime::3526 currently passes only by accident, `getEndOfVarScope()` is called with `tok` = `&` and variables `bar` and `_foo`. Since `Foo` lives in a namespace, `tok`'s scope is used, which is unrelated to the scope of either variable.

Reply:
E.g. [this](https://github.com/danmar/cppcheck/blob/main/test/testautovariables.cpp#L3562) breaks when another scope is added: `{ bar->_foo = &foo; }`